### PR TITLE
[B+C] API for Pathfinders. Adds BUKKIT-1710

### PR DESCRIPTION
**Pull request by tomsik68**
### The Issue

As all of us know, minecraft introduced a new API for mob Pathfinding. It was quite a long time ago and Bukkit is still missing support for Pathfinders. 
### Justification for this PR

We need an API for Pathfinders for a lots of awesome things. 
Like you can set Zombies not to break doors or you can have bunch of mobs following you,  change how monsters attack, change how fast mobs move, make mobs look at you.
### PR Breakdown
- Interface **Controllable**
  - Provides methods for looking at object, jumping and moving using vanilla jumpController, lookController and navigator from net.minecraft.server.Entity  
  - Methods to alter path finding like setAvoidWater, setBreakDoors, setAvoidSun which call vanilla methods from Navigator class.
- Interface **PathfindingEntity**
  - Pathfinder manipulation methods including clearPathfinders, which clears list a and b from goalSelector and targetSelector of EntityLiving and attachPathfinder, which uses PathfinderWrapper in org.bukkit.craftbukkit.ai to add to Entity's task list. Either goalSelector or targetSelector depending on what's specified in Pathfinder class.
  - **This interface is implemented by Animals and chosen types of other mobs**
- Interface **Pathfinder**
  - Pathfinder interface with few functions. This is whole new thing which plugin devs can use to change behavior of Entities.
  - Used by PathfinderWrapper
- Class **PathfinderWrapper**
  - Wrapper for Pathfinders
  - Extends minecraft's internal PathfinderGoal class and its field impl can store implemenetation of Pathfinder interface
  - When minecraft calls methods from it, methods from impl will be called
### Relevant PRs
- https://github.com/Bukkit/CraftBukkit/pull/1128 - My PR with the API created
### Testing material
- PathfinderTest - If you right-click a living entity it will start following you  
  - Plugin JAR - https://dl.dropbox.com/u/29628587/PathfinderTest.jar
  - Source - https://github.com/tomsik68/PathfinderTest
### JIRA Ticket
- https://bukkit.atlassian.net/browse/BUKKIT-1710 - I'm not creator of the ticket, but there's a link to this PR in comments

### CODE REVIEW

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * @author Tomsik68
+ *
+ */
+public interface Pathfinder {
+    /**
+     * 
+     * @return Entity controlled by the pathfinder
+     */
+    public LivingEntity getEntity();
+    /** Sets entity to be controlled by the pathfinder
```
Comment (andrepl) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: all of the javadocs are formatted improperly. should be:
    /**
     \* Description Goes Here
     */

you've got the description on the `/**` line.

Comment (tomsik68) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: Is it better now? :)

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * 
+     * @return Entity controlled by the pathfinder
+     */
+    public LivingEntity getEntity();
+    /**
+     * Sets entity to be controlled by the pathfinder
+     * @param entity
+     */
+    public void setEntity(LivingEntity entity);
```
Comment (andrepl) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: There should be a blank line between every method. as well as a blank line between the javadoc's first line and any others. eg:

```
/**
 * Sets entity to be controlled by the pathfinder
 *
 * @param entity
 */
```

the space before "Sets" is important, as is the blank line between, in fact the dev team might want a <p> there instead, someone else will know.

**Comment by natemort** [comment]
Your javadocs could use some improvement.  Additionally, we no longer use @author annotations.

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+     * 
+     * @return If pathfinder can start guiding the entity (if it can see the entity, if it's in love etc.)
+     */
+    public boolean canStart();
+    /**
+     * 
+     * @return If pathfinder should continue execution(if player is in required range etc.)
+     */
+    public boolean canContinue();
+    /** Called when pathfinder execution stops
```
Comment (andrepl) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: Missed one here

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+     * @param w
+     */
+    public void setAvoidWater(boolean w);
+    /**
+     * 
+     * @return If this entity can break doors (zombies in villages)
+     */
+    public boolean canBreakDoors();
+    /** 
+     *Sets if this entity can break doors
```
Comment (andrepl) [inline src/main/java/org/bukkit/entity/Intelligent.java]: - Sets
  ^ needs space

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+     * @param swim
+     */
+    public void setCanSwim(boolean swim);
+    /**
+     *Attaches a new {@link Pathfinder} to this entity.
+     * @param pf Pathfinder to attach.
+     */
+    public void attachPathfinder(Pathfinder pf);
+    /**
+     *Deletes all the pathfinders(even the vanilla ones)
```
Comment (andrepl) [inline src/main/java/org/bukkit/entity/Intelligent.java]: Space here too. \* Deletes

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public void setEntity(LivingEntity entity);
+    /**
+     * 
+     * @return If pathfinder can start guiding the entity (if it can see the entity, if it's in love etc.)
+     */
+    public boolean canStart();
+    /**
+     * 
+     * @return If pathfinder should continue execution(if player is in required range etc.)
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/Pathfinder.java:31]: How frequently is this checked?  When is this checked?

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    /** 
+     * Makes controlled entity look at specified place
+     * 
+     * @param x
+     * @param y
+     * @param z
+     * @param speedPitch
+     * @param speedYaw
+     */
+    public void lookAt(double x,double y,double z,float speedPitch,float speedYaw);
```
Comment (andrepl) [inline src/main/java/org/bukkit/entity/Intelligent.java]: spaces after commas, everywhere.

File: src/main/java/org/bukkit/ai/Pathfinder.java (hunk 3 above)
Comment (natemort) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: When does pathfinder execution stop?

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public PathfinderType getType();
+    /**
+     *Gets priority of this pathfinder. 0 = highest
+     * 
+     * @return Priority in range 0 - Integer.MAX_VALUE
+     */
+    public int getPriority();
+    /**
+     *Performs a logical update of this pathfinder.
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: How frequently is this called?  What should a pathfinder do on an update?

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public void onStop();
+    /** 
+     * Gets type of this Pathfinder. Can't be null.
+     * 
+     * @return
+     */
+    public PathfinderType getType();
+    /**
+     *Gets priority of this pathfinder. 0 = highest
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: What does the priority determine?

File: src/main/java/org/bukkit/ai/PathfinderType.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ */
+public enum PathfinderType {
+    /**
+     *Goal pathfinders are peaceful and help mobs to swim, panic and lots of more useful things
+     */
+    Goal,
+    /**
+     *Targetting pathfinders help monsters & golems to choose targets
+     *  Note: You should call {@link EntityTargetEvent} when you choose target
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/PathfinderType.java:15]: Requiring a plugin to throw the event feels a bit ugly to me.  Could this instead by handled by the implementation?

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Deletes all the pathfinders(even the vanilla ones)
+     */
+    public void clearPathfinders();
+    /** 
+     * Makes controlled entity look at another entity 
+     * 
+     * @param entity
+     * @param speedPitch
+     * @param speedYaw
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/Intelligent.java]: What do these parameters do?

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    public void lookAt(Entity entity,float speedPitch,float speedYaw);
+    /** 
+     * Makes controlled entity look at specified place
+     * 
+     * @param x
+     * @param y
+     * @param z
+     * @param speedPitch
+     * @param speedYaw
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/Intelligent.java]: What do these parameters do?

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    public void lookAt(double x,double y,double z,float speedPitch,float speedYaw);
+    /** 
+     * Moves entity to specified place with specified speed
+     * 
+     * @param x
+     * @param y
+     * @param z
+     * @param speed
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/Intelligent.java]: What do these parameters do?

**Comment by tomsik68** [comment]
@evilmidget38 What improvement? Formatting?
@andrepl What else do I need to improve?

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    public void moveTo(double x,double y,double z,float speed);
+    /** 
+     * Makes entity jump
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/Intelligent.java]: How high is a jump?  Is the velocity applied to the jump?  This needs a better explanation.

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
+package org.bukkit.ai;
+
+
+import org.bukkit.entity.LivingEntity;
+
+/** 
+ * Represents a Pathfinder for {@link Controllable}s
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: Where's this Controllable interface?

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+     * @return If pathfinder can start guiding the entity (if it can see the entity, if it's in love etc.)
+     */
+    public boolean canStart();
+    /**
+     * 
+     * @return If pathfinder should continue execution(if player is in required range etc.)
+     */
+    public boolean canContinue();
+    /** 
+     * Called when pathfinder execution stops
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: When does the pathfinder execution stop?  What determines this?

File: src/main/java/org/bukkit/ai/Pathfinder.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+     * 
+     */
+    public void onStop();
+    /** 
+     * Gets type of this Pathfinder. Can't be null.
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: Perhaps a brief explanation regarding which type should be returned would be appropriate here.

File: src/main/java/org/bukkit/ai/PathfinderType.java [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+import org.bukkit.event.entity.EntityTargetEvent;
+/** 
+ * Represents type of a Pathfinder
+ *
+ */
+public enum PathfinderType {
+    /**
+     * Goal pathfinders are peaceful and help mobs to swim, panic and lots of more useful things
+     */
+    Goal,
```
Comment (natemort) [inline src/main/java/org/bukkit/ai/PathfinderType.java]: enums should be all uppercase.

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+import org.bukkit.ai.Pathfinder;
+
+/**
+ * Represents an intelligent mob(every living entity with pathfinder)
+ * 
+ *
+ */
+public interface Intelligent {
+    /**
+     *
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/Intelligent.java]: Methods need a method description as well as a return/parameters description.

**Comment by andrepl** [comment]
@tomsik68 there are still a lot of whitespace issues, try to work through the ones that evilmidget38 and I pointed out and then we'll find some more :)  I went through the same thing with my first PR not very long ago, its a bit discouraging, but it's all worth it if the code gets accepted! :)

in general, you need to put blank lines between every method, and remove blank lines at the end of a class after the last method.  your if statements are all cramped like this:

```
if(this=that){
    dothis()
}else{
```

these should be:

```
if (this == that) {
    dothis()
} else {
```

**Comment by SXRWahrheit** [comment]
Evilmidget has left you a lot of helpful comments. :)

File: src/main/java/org/bukkit/entity/Intelligent.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    public void setEnterDoors(boolean pass);
+    /**
+     * Sets if this entity should avoid sun
+     * 
+     * @param noSun
+     */
+    public void setAvoidSun(boolean noSun);
+    /**
+     * Sets speed of this entity while moving. 1.0 is really fast
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/Intelligent.java]: How fast is really fast?

File: src/main/java/org/bukkit/entity/LivingEntity.java [hunk 21]
```diff
@@ -13,7 +13,7 @@
 /**
  * Represents a living entity, such as a monster or player
  */
-public interface LivingEntity extends Entity, Damageable {
+public interface LivingEntity extends Entity, Damageable, Intelligent {
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/LivingEntity.java]: Is this API applicable to truly all LivingEntities?  Is Intelligent the best name for the interface?  I feel like Controllable would make more sense, and that this interface shouldn't be implemented by all Entities(namely Players).

File: src/main/java/org/bukkit/ai/Pathfinder.java (hunk 15 above)
Comment (andrepl) [inline src/main/java/org/bukkit/ai/Pathfinder.java]: is this supposed to be 'Intelligent' ?

File: src/main/java/org/bukkit/entity/LivingEntity.java (hunk 21 above)
Comment (tomsik68) [inline src/main/java/org/bukkit/entity/LivingEntity.java]: Ok, I'll rename it back to Controllable. 
I based it on vanilla minecraft, as even players have those controllers there. I'm working on a change...

**Comment by andrepl** [comment]
Does this have any effect on mobs that don't use the new AI?  if not, it might be best to add the interface only to those entities that do.

**Comment by tomsik68** [comment]
@andrepl No, it doesn't. I'll do what you say and add it just to entities that can use pathfinders. Thanks for pointing that out ;)

File: src/main/java/org/bukkit/entity/Skeleton.java [hunk 22]
```diff
@@ -3,7 +3,7 @@
 /**
  * Represents a Skeleton.
  */
-public interface Skeleton extends Monster {
+public interface Skeleton extends Monster, PathfindingEntity {
```
Comment (natemort) [inline src/main/java/org/bukkit/entity/Skeleton.java:6]: Perhaps Monster should extend PathfindingEntity if all of its subinterfaces do.

**Comment by riking** [comment]
This seems to me like a pretty thin wrapper around the NMS methods.

Also, you need some kind of way to restore the default behavior once the plugin is done controlling the entity.

File: src/main/java/org/bukkit/ai/PathfinderType.java [hunk 23]
```diff
@@ -0,0 +1,18 @@
+package org.bukkit.ai;
+
+import org.bukkit.event.entity.EntityTargetEvent;
+/** 
+ * Represents type of a Pathfinder
+ *
+ */
+public enum PathfinderType {
+    /**
+     * Goal pathfinders are peaceful and help mobs to swim, panic and lots of more useful things
```
Comment (riking) [inline src/main/java/org/bukkit/ai/PathfinderType.java:10]: > Goal pathfinders are not directed at a specific entity. They may be targeted towards a certain block, but not a specific entity. Examples of goal pathfinders would be idling, swimming, panicking, a random stroll, looking at a player, and a sheep eating grass. Note that vanilla treats breeding as a goal pathfinder and not a target pathfinder.

File: src/main/java/org/bukkit/ai/PathfinderType.java [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+     */
+    GOAL,
+    /**
+     * Targetting pathfinders help monsters & golems to choose targets
```
Comment (riking) [inline src/main/java/org/bukkit/ai/PathfinderType.java:14]: > Target pathfinders are directed at a specific entity. Examples of target pathfinders would be monsters attacking a player, or a wolf attacking sheep.
> When you choose a target, you MUST call a {@link EntityTargetEvent}.

File: src/main/java/org/bukkit/entity/PathfindingEntity.java [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ */
+public interface PathfindingEntity extends Controllable, LivingEntity {
+    /**
+     * Attaches a new {@link Pathfinder} to this entity.
+     * @param pf - Pathfinder to attach.
+     */
+    public void attachPathfinder(Pathfinder pf);
+    /**
+     * Deletes all the pathfinders(even the vanilla ones)
```
Comment (riking) [inline src/main/java/org/bukkit/entity/PathfindingEntity.java:15]: > Clear all pathfinders from this entity. You are required to immediately attach additional pathfinders.

#### Create a new method to restore the vanilla pathfinders, please.

File: src/main/java/org/bukkit/entity/Controllable.java [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+     */
+    public void lookAt(Entity entity, float speedPitch, float speedYaw);
+    /**
+     * Makes controlled entity look at specified location
+     * @param location - Location where to look at
+     * @param speedPitch - Speed of rotation around the X-axis
+     * @param speedYaw - Speed of rotation around the Y-axis
+     */
+    public void lookAt(Location location, float speedPitch, float speedYaw);
+    /**
```
Comment (riking) [inline src/main/java/org/bukkit/entity/Controllable.java:70]: You need an extra linebreak between the method and the next javadoc.

Don't think that's documented anywhere, but do it.

**Comment by riking** [comment]
Please create a new method to restore the vanilla pathfinders.

Also, I think that you need to rethink the way you're implementing this. Write it out on paper, and stuff.

File: src/main/java/org/bukkit/entity/Skeleton.java (hunk 22 above)
Comment (tomsik68) [inline src/main/java/org/bukkit/entity/Skeleton.java:6]: Well, they don't... Blaze, EnderDragon, Enderman and lots of others don't...

**Comment by tomsik68** [comment]
@riking Ok, it seems like you're right as it's too thin for bukkit. I'll rethink a way of implementing and probably submit a new PR as this is already too much commits. Thanks everyone for helping out, I'm now writing new code...

**Pull request closed without merging**